### PR TITLE
ref(redis): Wrap redis cluster initialization in a SLO

### PR DESCRIPTION
**Pull request by evanpurkhiser**
Do not make connections on application boot.

### CODE REVIEW

**Comment by EvanPurkhiser** [comment]
I'm still understanding the behavior of this in the application itself. Right now the similar API endpoint just hangs if it fails to connect, whereas I would expect a 500.

**Comment by mattrobenolt** [comment]
> Right now the similar API endpoint just hangs if it fails to connect, whereas I would expect a 500.

It is possible that that's the behavior coming out of `RetryingStrictRedisCluster` now when it's trying to reconnect and looping? Since now it'd be triggering this behavior on the request, and not boot.

File: src/sentry/utils/redis.py [hunk 1]
```diff
@@ -103,13 +103,12 @@ def factory(self, **config):
...
-        try:
+        # Redis cluster does not wait to attempt to connect. We'd prefer to not
+        # make TCP connections on boot. Wrap the client in a lazy proxy object.
+        def cluster_factory():
             return RetryingStrictRedisCluster(startup_nodes=hosts, decode_responses=True)
-        except RedisClusterException:
-            logger.warning('Failed to connect to Redis Cluster', exc_info=True)
-            raise KeyError('Redis Cluster could not be initalized')
+
+        return SimpleLazyObject(cluster_factory)
```
Comment (mattrobenolt) [inline src/sentry/utils/redis.py]: You could have just used a lambda. :)

```python
return SimpleLazyObject(lambda: RetryingStrictRedisCluster(startup_nodes=hosts, decode_responses=True))
```

Comment (EvanPurkhiser) [reply src/sentry/utils/redis.py]: I considered that, but I liked the named function a little bit more. I can switch it over if you think it falls more in line without our satndards.

Comment (mattrobenolt) [reply src/sentry/utils/redis.py]: Nah, I wasn't really suggesting to change anything. Just my preference to use a lambda for this, but that's my subjective opinion.

**Comment by EvanPurkhiser** [comment]
> It is possible that that's the behavior coming out of RetryingStrictRedisCluster now when it's trying to reconnect and looping? Since now it'd be triggering this behavior on the request, and not boot.

It's not this, I see the exception in the console. It bubbles up here, but for whatever reason, the response is not returned.

https://github.com/getsentry/sentry/blob/01d4819d83d82b3f62712c9d9a796526fdd76ee1/src/sentry/api/base.py#L80-L92

**Comment by EvanPurkhiser** [comment]
It looks like `raven.captureException` is eating it and throwing something itself, which is never logged into the console. I'm surprised though that the request is never completed.

I'm not going to dig into that anymore since I'm sure that's just something wonky about my dev setup. This is good to go.

**Comment by dcramer** [comment]
Is there a reason that using redis in a standard way used throughout other services via ``LazyServiceWrapper`` doesnt solve this?

**Comment by EvanPurkhiser** [comment]
The `rb` client doesn't make a connection to the server until it's first used, the `StrictRedisCluster` however, *does*. We're trying to avoid this kind of initialization during application boot.

**Comment by tkaemming** [comment]
> I'm not going to dig into that anymore since I'm sure that's just something wonky about my dev setup. This is good to go.

I also encounter this locally. Specifically,

```
20:20:08 server  | Traceback (most recent call last):
20:20:08 server  |   File "/Users/ted/workspaces/sentry/sentry/src/sentry/api/base.py", line 82, in handle_exception
20:20:08 server  |     return super(Endpoint, self).handle_exception(exc)
20:20:08 server  |   File "/Users/ted/workspaces/sentry/sentry/src/sentry/api/base.py", line 177, in dispatch
20:20:08 server  |     response = handler(request, *args, **kwargs)
20:20:08 server  |   File "/Users/ted/workspaces/sentry/sentry/src/sentry/api/endpoints/group_similar_issues.py", line 22, in get
20:20:08 server  |     features.compare(group, limit=limit)
20:20:08 server  |   File "/Users/ted/workspaces/sentry/sentry/src/sentry/similarity/features.py", line 212, in compare
20:20:08 server  |     limit=limit,
20:20:08 server  |   File "/Users/ted/workspaces/sentry/sentry/src/sentry/similarity/backends/metrics.py", line 27, in compare
20:20:08 server  |     return self.__instrumented_method_call('compare', *args, **kwargs)
20:20:08 server  |   File "/Users/ted/workspaces/sentry/sentry/src/sentry/similarity/backends/metrics.py", line 18, in __instrumented_method_call
20:20:08 server  |     return getattr(self.backend, method)(scope, *args, **kwargs)
20:20:08 server  |   File "/Users/ted/workspaces/sentry/sentry/src/sentry/similarity/backends/redis.py", line 127, in compare
20:20:08 server  |     return self._as_search_result(self.__index(scope, arguments))
20:20:08 server  |   File "/Users/ted/workspaces/sentry/sentry/src/sentry/similarity/backends/redis.py", line 48, in __index
20:20:08 server  |     return index(self.cluster, [scope], args)
20:20:08 server  |   File "/Users/ted/workspaces/sentry/sentry/src/sentry/utils/redis.py", line 231, in call_script
20:20:08 server  |     return script(keys, args, client)
20:20:08 server  |   File "/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/redis/client.py", line 2694, in __call__
20:20:08 server  |     return client.evalsha(self.sha, len(keys), *args)
20:20:08 server  |   File "/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/django/utils/functional.py", line 213, in inner
20:20:08 server  |     self._setup()
20:20:08 server  |   File "/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/django/utils/functional.py", line 298, in _setup
20:20:08 server  |     self._wrapped = self._setupfunc()
20:20:08 server  |   File "/Users/ted/workspaces/sentry/sentry/src/sentry/utils/redis.py", line 109, in cluster_factory
20:20:08 server  |     return RetryingStrictRedisCluster(startup_nodes=hosts, decode_responses=True)
20:20:08 server  |   File "/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/rediscluster/client.py", line 181, in __init__
20:20:08 server  |     **kwargs
20:20:08 server  |   File "/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/rediscluster/connection.py", line 141, in __init__
20:20:08 server  |     self.nodes.initialize()
20:20:08 server  |   File "/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/rediscluster/nodemanager.py", line 240, in initialize
20:20:08 server  |     raise RedisClusterException("Redis Cluster cannot be connected. Please provide at least one reachable node.")
20:20:08 server  | RedisClusterException: Redis Cluster cannot be connected. Please provide at least one reachable node.
20:20:10 server  | Traceback (most recent call last):
20:20:10 server  |   File "/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/raven/utils/serializer/manager.py", line 76, in transform
20:20:10 server  |     return repr(value)
20:20:10 server  |   File "/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/rediscluster/client.py", line 219, in __repr__
20:20:10 server  |     servers = list({'{0}:{1}'.format(nativestr(info['host']), info['port']) for info in self.connection_pool.nodes.startup_nodes})
20:20:10 server  | AttributeError: 'RetryingStrictRedisCluster' object has no attribute 'connection_pool'
20:20:10 server  | 20:20:10 [ERROR] sentry.errors.serializer: 'RetryingStrictRedisCluster' object has no attribute 'connection_pool'
```

**Review by tkaemming — APPROVED** [review]
I don't think I've heard anyone refer to `SimpleLazyObject` as "SLO" but otherwise this looks good to me.

**Pull request merged**